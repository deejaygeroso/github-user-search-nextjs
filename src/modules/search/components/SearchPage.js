import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import { SpinLoader } from "react-css-loaders";
import Router from 'next/router'
import Typography from "@material-ui/core/Typography";

import SearchField from "./SearchField";
import UserCard from "./UserCard";
import PaginationCard from "./PaginationCard";

const styles = theme => ({
  root: {
    marginRight: 20,
    marginLeft: 20,
    paddingTop: theme.spacing.unit * 10
  },
  progress: {
    margin: theme.spacing.unit * 2
  },
  logo: {
    height: 80
  }
});

class SearchPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      username: "",
      page: 1,
      offset: ""
    };
  }

  /* ----------------------------------------------------------------------------------
   * After mounting get username & page parameter values on url if exist then
   * fetch resulting data from github api. 
   * -------------------------------------------------------------------------------- */
  componentDidMount(){
    const { userActions } = this.props;
    // used for getting url params
    var vars = {};
    window.location.href.replace(/[?&]+([^=&]+)=([^&]*)/gi, function(m,key,value) {
        vars[key] = value;
    });
    if(vars.page){
      this.setState({page: parseInt(vars.page)})
    }
    if(vars.username){
      this.setState({username: vars.username})
      userActions.apiGithubSearchUsers({ username: vars.username, page: vars.page });
    }
  }
  
  /* ----------------------------------------------------------------------------------
   * For Listing Users and showing loading when data is still being fetched 
   * -------------------------------------------------------------------------------- */
  reunderUserCard() {
    const { userList, userRequestStatus } = this.props;
    // show loading css
    if (userRequestStatus.isFetching) {
      return <SpinLoader />;
    }
    return (
      <Fragment>
        {userList.allIds.length !== 0 &&
          userList.allIds.map(userId => (
            <UserCard key={userId} user={userList.byId[userId]} onClick={()=>this.navigateToUser(userList.byId[userId])}/>
          ))}
      </Fragment>
    );
  }

  /* ----------------------------------------------------------------------------------
   * Main Component. 
   * -------------------------------------------------------------------------------- */
  render() {
    const { classes, userList, userRequestStatus } = this.props;
    return (
      <div className={classes.root}>
        <img className={classes.logo} src="/static/assets/images/logo.jpg" alt="Github User Search"/>
        <Typography variant="h4" gutterBottom>
          Github User Search
        </Typography>
        <SearchField onSubmit={this.onSubmit} onChange={this.onSearchChange} />
        {this.reunderUserCard()}
        <br />
        <PaginationCard
          isHidden={userRequestStatus.isFetching}
          offset={(userList.page-1) * 20}
          total={userList.total_count}
          onClick={this.onPaginate}
        />
      </div>
    );
  }

  /* ----------------------------------------------------------------------------------
   * Navigate to user page. 
   * -------------------------------------------------------------------------------- */
  navigateToUser = (user) => {
    const { userActions } = this.props;
    userActions.apiGithubUserAdditionalInfo({username: user.login});
    Router.push({
      pathname: '/user',
      query: { username: user.login }
    })
  }

  /* ----------------------------------------------------------------------------------
   * While typing update username state. 
   * -------------------------------------------------------------------------------- */
  onSearchChange = username => {
    this.setState({ username });
  };

  /* ----------------------------------------------------------------------------------
   * Find user by username. 
   * -------------------------------------------------------------------------------- */
  onSubmit = username => {
    const { userActions } = this.props;
    this.setState({ username });
    userActions.apiGithubSearchUsers({ username });
    const href = `/?username=${username}`;
    const as = href;
    Router.replace(href, as, { shallow: true })
  };

  /* ----------------------------------------------------------------------------------
   * Find user by username with offset. 
   * -------------------------------------------------------------------------------- */
  onPaginate = page => {
    const { username } = this.state;
    const { userActions } = this.props;
    userActions.apiGithubSearchUsers({ username, page });
    const href = `/?username=${username}&page=${page}`;
    const as = href;
    Router.push(href, as, { shallow: true })
  };
}

SearchPage.propTypes = {
  userList: PropTypes.object.isRequired,
  userActions: PropTypes.object.isRequired,
  userRequestStatus: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired
};

export default withStyles(styles)(SearchPage);
