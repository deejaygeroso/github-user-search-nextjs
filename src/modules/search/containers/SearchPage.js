import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import SearchPage from "../components/SearchPage";

import * as userActions from "../../../actions/userActions";

export const mapStateToProps = state => ({
  userList: state.userList,
  userRequestStatus: state.userRequestStatus
});

export const mapDispatchToProps = dispatch => ({
  userActions: bindActionCreators(userActions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchPage);
