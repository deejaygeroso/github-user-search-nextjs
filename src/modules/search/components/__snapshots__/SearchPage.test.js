import React from "react";
import SearchPage from "../SearchPage";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const userList = {
    allIds: [1, 2],
    byId: {
      1: {
        id: 1,
        login: "username 1",
        followers: 650,
        following: 12
      },
      2: {
        id: 2,
        login: "username 2",
        followers: 63,
        following: 24
      }
    },
    page: 1,
    total_count: 2,
  };
  const userRequestStatus = { isFetching: false };
  const userActions = {
    apiGithubSearchUsers : () => {},
    apiGithubUserAdditionalInfo : () => {}
  }
  const tree = renderer
    .create(
      <SearchPage
        userList={userList}
        userRequestStatus={userRequestStatus}
        userActions={userActions}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
