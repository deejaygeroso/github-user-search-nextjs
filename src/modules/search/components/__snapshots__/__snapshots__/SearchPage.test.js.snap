// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="SearchPage-root-1"
>
  <img
    alt="Github User Search"
    className="SearchPage-logo-3"
    src="/static/assets/images/logo.jpg"
  />
  <h4
    className="MuiTypography-root-4 MuiTypography-h4-19"
  >
    Search
  </h4>
  <p
    className="MuiTypography-root-4 MuiTypography-body1-13 MuiTypography-gutterBottom-31"
  >
    Github Users Here!
  </p>
  <div
    className="MuiPaper-root-43 MuiPaper-elevation1-46 MuiPaper-rounded-44 SearchField-root-40"
  >
    <div
      className="MuiInputBase-root-70 SearchField-input-41"
      onClick={[Function]}
      onKeyPress={[Function]}
    >
      <input
        className="MuiInputBase-input-80"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search by username"
        type="text"
      />
    </div>
    <button
      aria-label="Search"
      className="MuiButtonBase-root-93 MuiIconButton-root-87 SearchField-iconButton-42"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiIconButton-label-92"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-96"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-120"
      />
    </button>
  </div>
  <div
    className="MuiPaper-root-43 MuiPaper-elevation1-46 MuiPaper-rounded-44 MuiCard-root-107 UserCard-card-105"
  >
    <button
      className="MuiButtonBase-root-93 MuiCardActionArea-root-108"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <div
        className="MuiCardHeader-root-111"
      >
        <div
          className="MuiCardHeader-avatar-112"
        >
          <div
            aria-label="Recipe"
            className="MuiAvatar-root-117 MuiAvatar-colorDefault-118 UserCard-avatar-106"
          />
        </div>
        <div
          className="MuiCardHeader-content-114"
        >
          <span
            className="MuiTypography-root-4 MuiTypography-body2-12 MuiCardHeader-title-115"
          >
            username 1
          </span>
          <span
            className="MuiTypography-root-4 MuiTypography-body2-12 MuiTypography-colorTextSecondary-37 MuiCardHeader-subheader-116"
          >
            Followers: 650 Following: 12
          </span>
        </div>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight-110"
      />
      <span
        className="MuiTouchRipple-root-120"
      />
    </button>
  </div>
  <div
    className="MuiPaper-root-43 MuiPaper-elevation1-46 MuiPaper-rounded-44 MuiCard-root-107 UserCard-card-105"
  >
    <button
      className="MuiButtonBase-root-93 MuiCardActionArea-root-108"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <div
        className="MuiCardHeader-root-111"
      >
        <div
          className="MuiCardHeader-avatar-112"
        >
          <div
            aria-label="Recipe"
            className="MuiAvatar-root-117 MuiAvatar-colorDefault-118 UserCard-avatar-106"
          />
        </div>
        <div
          className="MuiCardHeader-content-114"
        >
          <span
            className="MuiTypography-root-4 MuiTypography-body2-12 MuiCardHeader-title-115"
          >
            username 2
          </span>
          <span
            className="MuiTypography-root-4 MuiTypography-body2-12 MuiTypography-colorTextSecondary-37 MuiCardHeader-subheader-116"
          >
            Followers: 63 Following: 24
          </span>
        </div>
      </div>
      <span
        className="MuiCardActionArea-focusHighlight-110"
      />
      <span
        className="MuiTouchRipple-root-120"
      />
    </button>
  </div>
  <br />
  <div />
</div>
`;
